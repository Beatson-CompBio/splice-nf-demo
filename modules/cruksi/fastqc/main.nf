process FASTQC {
  tag { "${meta.id}" }
  publishDir { params.outdir ? params.outdir + "/qc/fastqc" : "${projectDir}/results/qc/fastqc" }, mode: params.publish_dir_mode
  cpus 1
  memory '1 GB'
  time '2h'

  // Packaging options
  conda "${moduleDir}/environment.yml"
  container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
      'https://depot.galaxyproject.org/singularity/fastqc:0.12.1--hdfd78af_0' :
      'biocontainers/fastqc:0.12.1--hdfd78af_0' }"

  input:
    tuple val(meta), path(reads)

  output:
    tuple val(meta), path("*.html"), emit: html
    tuple val(meta), path("*.zip"),  emit: zip

  when:
    !params.skip_fastqc

  script:
  def extra = ''
  """
  fastqc --quiet --outdir . ${reads}
  """
}
