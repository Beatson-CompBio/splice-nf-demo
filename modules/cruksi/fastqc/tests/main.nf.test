nextflow_process {

    name "Test Process FASTQC"
    script "../main.nf"
    process "FASTQC"

    tag "modules"
    tag "modules_cruksi"
    tag "fastqc"

    test("fastqc - single end") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:true ],
                    [ file('https://raw.githubusercontent.com/nf-core/test-datasets/viralrecon/illumina/amplicon/sample1_R1.fastq.gz', checkIfExists: true) ]
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.html.size() == 1 },
                { assert process.out.zip.size() == 1 },
                { assert process.out.versions.size() == 1 },
                // Test file existence and properties instead of exact checksums
                { assert path(process.out.html.get(0).get(1)).exists() },
                { assert path(process.out.html.get(0).get(1)).size() > 1000 }, // HTML should be > 1KB
                { assert path(process.out.zip.get(0).get(1)).exists() },
                { assert path(process.out.zip.get(0).get(1)).size() > 1000 }, // ZIP should be > 1KB
                // Only snapshot the versions file (most stable across environments)
                { assert snapshot(process.out.versions).match("versions_single") }
            )
        }
    }

    test("fastqc - paired end") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ],
                    [ 
                        file('https://raw.githubusercontent.com/nf-core/test-datasets/viralrecon/illumina/amplicon/sample1_R1.fastq.gz', checkIfExists: true),
                        file('https://raw.githubusercontent.com/nf-core/test-datasets/viralrecon/illumina/amplicon/sample1_R2.fastq.gz', checkIfExists: true)
                    ]
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.html.size() == 1 },
                { assert process.out.zip.size() == 1 },
                { assert process.out.versions.size() == 1 },
                // Test that we get 2 HTML files for paired-end
                { assert process.out.html.get(0).get(1).size() == 2 },
                { assert process.out.zip.get(0).get(1).size() == 2 },
                // Check files exist and have reasonable sizes
                { assert path(process.out.html.get(0).get(1)[0]).exists() },
                { assert path(process.out.html.get(0).get(1)[1]).exists() },
                { assert path(process.out.html.get(0).get(1)[0]).size() > 1000 },
                { assert path(process.out.html.get(0).get(1)[1]).size() > 1000 },
                // Only snapshot the versions file
                { assert snapshot(process.out.versions).match("versions_paired") }
            )
        }
    }

    test("fastqc - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ],
                    [ 
                        file('https://raw.githubusercontent.com/nf-core/test-datasets/viralrecon/illumina/amplicon/sample1_R1.fastq.gz', checkIfExists: true),
                        file('https://raw.githubusercontent.com/nf-core/test-datasets/viralrecon/illumina/amplicon/sample1_R2.fastq.gz', checkIfExists: true)
                    ]
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.html.size() == 1 },
                { assert process.out.zip.size() == 1 },
                { assert process.out.versions.size() == 1 },
                // For stub, just check files exist
                { assert path(process.out.html.get(0).get(1)).exists() },
                { assert path(process.out.zip.get(0).get(1)).exists() }
            )
        }
    }
}
