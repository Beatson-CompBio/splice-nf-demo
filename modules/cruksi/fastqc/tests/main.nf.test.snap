{
    "fastqc - paired end": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    [
                        "sample1_R1_fastqc.html:md5,41b677f38b0f86944d035393a1353623",
                        "sample1_R2_fastqc.html:md5,e5f81fbe9ebbe65540963a47209fb05b"
                    ]
                ]
            ],
            [
                "versions.yml:md5,e1cc25ca8af856014824abd842e93978"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-09-02T09:05:51.186408663"
    },
    "fastqc - single end": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "sample1_R1_fastqc.html:md5,41b677f38b0f86944d035393a1353623"
                ]
            ],
            [
                "versions.yml:md5,e1cc25ca8af856014824abd842e93978"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-09-02T09:05:38.634737475"
    }
}